test/integration/xds_integration_test.cc

```diff
@@ -628,11 +628,12 @@ TEST_P(LdsInplaceUpdateTcpProxyIntegrationTest,
   while (!client_conn_0->closed()) {
     dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
   }
+  // Wait for the filter chain removal start. Ideally we have `drain_time_` to detect the
+  // value 1. Increase the drain_time_ at the beginning of the test if the test is flaky.
+  test_server_->waitForGaugeEq("listener_manager.total_filter_chains_draining", 1);
```

Comment:
I think this is flaky.  The drain could have completed by the time you attempt to check for the count being 1; this call will fail due to timeout.

Reply:
I agree. This might be flaky in CI. We cannot inject the wait earlier because we wants to stack two listener update as closer as possible.

At the cost of flakines, this test case crashes less than 1 out of 100 under buggy production code (connection_handler_impl.cc before this PR).
The manipulated order inversion is explicit in unitest test connection_handler_impl_test so I think we should delete this integration test.

Reply:
Can you use simulated time?

Reply:
> Can you use simulated time?

The bug was raised because 2 timers at the same time expiration are in the order that timers are inserted. 

SimulatedTimeSystem can move *wall time* back but the timer expiration provided by SimulatedTimeSystem is based on monotonic time which is designed to be forward only. So I cannot use SimulatedTimeSystem to reproduce the desired out-of-order stably.

Reply:
Is there a way to improve STS to make this possible to model deterministically? CC @jmarantz

Reply:
To produce the crash in real world, at the time point, two timer expirations are set at X seconds later, and the second is triggered prior to the first one.

I tried to move backward the monotonic time before I enable the second timer. Yet it's not deterministic.
https://github.com/envoyproxy/envoy/blob/main/test/test_common/simulated_time_system.h#L38-L46
```
  /**
   * Sets the time forward monotonically. If the supplied argument moves
   * backward in time, the call is a no-op...
```

The good part of the integration test case is that it's not flaky. It passes 100%.
The bad is that this test passes 99% on the buggy code...

I am wondering if the unit test is good enough and we delete this integration test case since it's a low signal.

Reply:
So with correct code, --runs_per_test=1000 passes, and with incorrect code, --runs_per_test=1000 fails ~10 times?

Can we be happy with that for validating a fix?

Reply:
> So with correct code, --runs_per_test=1000 passes, and with incorrect code, --runs_per_test=1000 fails ~10 times?
> 
> Can we be happy with that for validating a fix?

Yes. I am confident about the fix. Could we land this PR? So we can verify it in istio CI and detect further crash source

Reply:
I can create a issue and make the test crash harder at incorrect code with @antoniovicente

Reply:
@avd and @htuch are the gatekeepers on this one but it's OK with me!

Reply:
```
$ bazel test //test/integration:xds_integration_test --test_filter="*TcpListenerRemoveFilterChainCalledAfterListenerIsRemoved*" --test_arg="-l debug" --runs_per_test=100
//test/integration:xds_integration_test                                  FAILED in 47 out of 100 in 2.5s
```